From dcf8a17219e4b8cca2a2dc92635d4ef9e3045458 Mon Sep 17 00:00:00 2001
From: Stanislav Asunkin <1353637+stasjok@users.noreply.github.com>
Date: Thu, 23 Oct 2025 13:57:53 +0300
Subject: perf: improve symbol-extractors performance

* combine all queries in one file
* extract date using captures
* use one function and one loop to collect all matches

diff --git a/packages/lsp-server/src/common/common.ts b/packages/lsp-server/src/common/common.ts
index 0d8f44e..e1d802a 100644
--- a/packages/lsp-server/src/common/common.ts
+++ b/packages/lsp-server/src/common/common.ts
@@ -74,6 +74,14 @@ export function rangeToCompact(range: lsp.Range): [number, number, number, numbe
 	return [range.start.line, range.start.character, range.end.line, range.end.character];
 }
 
+/**
+ * Converts a Tree-sitter node range to a compact array representation [startLine, startChar, endLine, endChar]
+ * This saves memory when storing many ranges in a database
+ */
+export function nodeToCompact(node: Parser.SyntaxNode): [number, number, number, number] {
+	return [node.startPosition.row, node.startPosition.column, node.endPosition.row, node.endPosition.column];
+}
+
 /**
  * Converts a compact range representation back to an LSP Range object
  */
diff --git a/packages/lsp-server/src/common/features/symbol-extractors.ts b/packages/lsp-server/src/common/features/symbol-extractors.ts
index 6642182..113d79a 100644
--- a/packages/lsp-server/src/common/features/symbol-extractors.ts
+++ b/packages/lsp-server/src/common/features/symbol-extractors.ts
@@ -1,7 +1,7 @@
 import * as lsp from 'vscode-languageserver';
 import { TextDocument } from 'vscode-languageserver-textdocument';
 import type { SyntaxNode } from 'web-tree-sitter';
-import { asLspRange, compactToRange, rangeToCompact } from '../common';
+import { asLspRange, compactToRange, nodeToCompact, rangeToCompact } from '../common';
 import { TreeQuery } from '../language';
 import { Trees } from '../trees';
 
@@ -94,6 +94,110 @@ function findDateFromNode(
 	return undefined;
 }
 
+function stripFirstChar(s: string): string {
+	return s.substring(1)
+}
+
+export async function getSymbols(textDocument: TextDocument, trees: Trees): Promise<SymbolInfo[]> {
+	const tree = await trees.getParseTree(textDocument);
+	if (!tree) {
+		throw new Error(`Failed to get parse tree for document: ${textDocument.uri}`);
+	}
+	const query = TreeQuery.getQueryByTokenName('symbols');
+	const matches = await query.matches(tree.rootNode);
+	const result: SymbolInfo[] = [];
+	for (const match of matches) {
+		let nameFromCapture: string | undefined
+		let nameMod: ((name: string) => string) | undefined
+		let date: string | undefined
+		let symbolInfos: SymbolInfo[] = []
+		for (const capture of match.captures) {
+			let type: typeof SymbolType[keyof typeof SymbolType]
+			let name: string | undefined
+			let kind: lsp.SymbolKind
+			switch (capture.name) {
+				case "currency":
+					type = SymbolType.COMMODITY
+					kind = lsp.SymbolKind.Constant
+					break
+				case "account_usage":
+					type = SymbolType.ACCOUNT_USAGE
+					kind = lsp.SymbolKind.Struct
+					break
+				case "date":
+					date = capture.node.text
+					continue;
+				case "narration":
+					type = SymbolType.NARRATION
+					name = capture.node.text.slice(1, -1)
+					kind = lsp.SymbolKind.String
+					break
+				case "payee":
+					type = SymbolType.PAYEE
+					name = capture.node.text.slice(1, -1)
+					kind = lsp.SymbolKind.String
+					break
+				case "tag":
+					type = SymbolType.TAG
+					name = capture.node.text.substring(1)
+					kind = lsp.SymbolKind.Key
+					break
+				case "link":
+					type = SymbolType.LINK
+					name = capture.node.text.substring(1)
+					kind = lsp.SymbolKind.Key
+					break
+				case "price":
+					type = SymbolType.PRICE
+					kind = lsp.SymbolKind.Constant
+					break
+				case "account_definition":
+					type = SymbolType.ACCOUNT_DEFINITION
+					kind = lsp.SymbolKind.Struct
+					break
+				case "close":
+					type = SymbolType.ACCOUNT_CLOSE
+					kind = lsp.SymbolKind.Struct
+					break
+				case "currency_definition":
+					type = SymbolType.CURRENCY_DEFINITION
+					kind = lsp.SymbolKind.Constant
+					break
+				case "pushtag":
+					type = SymbolType.PUSHTAG
+					nameMod = stripFirstChar
+					kind = lsp.SymbolKind.Event
+					break
+				case "poptag":
+					type = SymbolType.POPTAG
+					nameMod = stripFirstChar
+					kind = lsp.SymbolKind.Event
+					break
+				case "name":
+					nameFromCapture = capture.node.text
+					continue;
+				default:
+					continue;
+			}
+			symbolInfos.push({
+				[SymbolKey.TYPE]: type,
+				_uri: textDocument.uri,
+				name: name ?? capture.node.text,
+				range: nodeToCompact(capture.node),
+				kind: kind,
+			})
+		}
+		for (const symbolInfo of symbolInfos) {
+			symbolInfo.date = date
+			if (nameFromCapture) {
+				symbolInfo.name = nameMod ? nameMod(nameFromCapture) : nameFromCapture
+			}
+			result.push(symbolInfo)
+		}
+	}
+	return result
+}
+
 export async function getAccountsUsage(textDocument: TextDocument, trees: Trees): Promise<SymbolInfo[]> {
 	const query = TreeQuery.getQueryByTokenName('account_usage');
 	const tree = await trees.getParseTree(textDocument);
diff --git a/packages/lsp-server/src/common/features/symbol-index.ts b/packages/lsp-server/src/common/features/symbol-index.ts
index c69c256..6c96f9c 100644
--- a/packages/lsp-server/src/common/features/symbol-index.ts
+++ b/packages/lsp-server/src/common/features/symbol-index.ts
@@ -6,18 +6,7 @@ import { DocumentStore } from '../document-store';
 import { StorageInstance } from '../startServer';
 import { Trees } from '../trees';
 import {
-	getAccountsClose,
-	getAccountsDefinition,
-	getAccountsUsage,
-	getCommodities,
-	getCurrencyDefinitions,
-	getLinks,
-	getNarrations,
-	getPayees,
-	getPopTags,
-	getPricesDeclarations,
-	getPushTags,
-	getTags,
+	getSymbols,
 	SymbolInfo,
 	SymbolKey,
 	SymbolType,
@@ -197,62 +186,12 @@ export class SymbolIndex {
 		// Process options directives in this document
 		await this._processOptionsDirectives(document);
 
-		const [
-			accountUsages,
-			accountDefinitions,
-			accountCloseDertives,
-			payees,
-			narrations,
-			commodities,
-			tags,
-			pushTags,
-			popTags,
-			pricesDeclarations,
-			links,
-		] = await Promise.all(
-			[
-				getAccountsUsage(document, this._trees),
-				getAccountsDefinition(document, this._trees),
-				getAccountsClose(document, this._trees),
-				getPayees(document, this._trees),
-				getNarrations(document, this._trees),
-				getCommodities(document, this._trees),
-				getTags(document, this._trees),
-				getPushTags(document, this._trees),
-				getPopTags(document, this._trees),
-				getPricesDeclarations(document, this._trees),
-				getLinks(document, this._trees),
-			],
-		);
-
-		this.logger.debug(`We Found symbols in ${document.uri}:
-			- Account usages: ${accountUsages.length}
-			- Account definitions: ${accountDefinitions.length}
-			- Account closures: ${accountCloseDertives.length}
-			- Payees: ${payees.length}
-			- Narrations: ${narrations.length}
-			- Commodities: ${commodities.length}
-			- Tags: ${tags.length}
-			- Push tags: ${pushTags.length}
-			- Pop tags: ${popTags.length}
-			- Prices declarations: ${pricesDeclarations.length}
-			- Links: ${links.length}
-		`);
+		const symbols = await getSymbols(document, this._trees)
+
+		this.logger.debug(`We Found ${symbols.length} symbols in ${document.uri}`);
 
 		this._symbolInfoStorage.removeSync({ _uri: document.uri });
-		await Promise.all([
-			this._symbolInfoStorage.insertAsync(accountUsages),
-			this._symbolInfoStorage.insertAsync(accountDefinitions),
-			this._symbolInfoStorage.insertAsync(accountCloseDertives),
-			this._symbolInfoStorage.insertAsync(payees),
-			this._symbolInfoStorage.insertAsync(narrations),
-			this._symbolInfoStorage.insertAsync(commodities),
-			this._symbolInfoStorage.insertAsync(tags),
-			this._symbolInfoStorage.insertAsync(pushTags),
-			this._symbolInfoStorage.insertAsync(popTags),
-			this._symbolInfoStorage.insertAsync(pricesDeclarations),
-			this._symbolInfoStorage.insertAsync(links),
-		]);
+		await this._symbolInfoStorage.insertAsync(symbols)
 
 		await yieldToMain();
 		const tree = await this._trees.getParseTree(document);
@@ -280,12 +219,6 @@ export class SymbolIndex {
 				this.unleashFiles([]);
 			}
 		}
-
-		// Add currency definitions
-		const currencyDefinitions = await getCurrencyDefinitions(document, this._trees);
-		if (currencyDefinitions.length > 0) {
-			await this._symbolInfoStorage.insertAsync(currencyDefinitions);
-		}
 	}
 
 	/**
diff --git a/packages/lsp-server/src/common/language/index.ts b/packages/lsp-server/src/common/language/index.ts
index 810f902..751dc9e 100644
--- a/packages/lsp-server/src/common/language/index.ts
+++ b/packages/lsp-server/src/common/language/index.ts
@@ -37,6 +37,8 @@ import account_definition from './queries/account_definition.scm';
 import account_usage from './queries/account_usage.scm';
 import currency_definition from './queries/currency_definition.scm';
 
+import symbols from './queries/symbols.scm'
+
 const queryMap = {
 	account,
 	balance,
@@ -70,6 +72,7 @@ const queryMap = {
 	account_definition,
 	account_usage,
 	currency_definition,
+	symbols
 } as const;
 
 type BeanTokenName = keyof typeof queryMap;
diff --git a/packages/lsp-server/src/common/language/queries/symbols.scm b/packages/lsp-server/src/common/language/queries/symbols.scm
new file mode 100644
index 0000000..90b72e1
--- /dev/null
+++ b/packages/lsp-server/src/common/language/queries/symbols.scm
@@ -0,0 +1,53 @@
+(transaction
+  payee: (payee)? @payee
+  narration: (narration) @narration)
+
+(transaction
+  date: (date) @date
+  (posting 
+    account: (account) @account_usage))
+
+(pad
+  date: (date) @date
+  account: (account) @account_usage
+  from_account: (account) @account_usage)
+
+(balance 
+  date: (date) @date
+  account: (account) @account_usage)
+
+(close 
+  date: (date) @date
+  account: (account) @account_usage @name) @close
+
+(note 
+  date: (date) @date
+  account: (account) @account_usage)
+
+(document 
+  date: (date) @date
+  account: (account) @account_usage)
+
+(open 
+  date: (date) @date
+  account: (account) @account_definition)
+
+(commodity
+  date: (date) @date
+  currency: (currency) @currency_definition) 
+
+(currency) @currency
+
+(tag) @tag
+
+(link) @link
+
+(pushtag
+  (tag) @name) @pushtag
+
+(poptag
+  (tag) @name) @poptag
+
+(price
+  date: (date) @date
+  currency: (currency) @name) @price
-- 
2.50.1

